server/db/mysql/adapter.go

```diff
@@ -60,17 +72,33 @@ func (a *adapter) Open(jsonconfig json.RawMessage) error {
 	}
 
 	var err error
-	var config configType
+	defaultCfg := ms.NewConfig()
+	config := configType{Config: *defaultCfg}
 	if err = json.Unmarshal(jsonconfig, &config); err != nil {
 		return errors.New("mysql adapter failed to parse config: " + err.Error())
 	}
 
-	a.dsn = config.DSN
-	if a.dsn == "" {
-		a.dsn = defaultDSN
+	locationCmp := cmp.Comparer(func(x, y time.Location) bool {
+		return x.String() == y.String()
+	})
+	if !cmp.Equal(*defaultCfg, config.Config, locationCmp, cmpopts.IgnoreUnexported(ms.Config{})) {
```

Comment:
I don't understand this. `config.Config` is identical to `*defaultCfg`:
```go
config := configType{Config: *defaultCfg}
```
How can they ever be different?

Reply:
I think it might be easier to just call `config.FormatDSN()` and compare it to the DSN you get from the default config. If they are different then Config must have been provided. If they are the same then the Config is blank.

Reply:
That sounds reasonable. Changed.